### PR TITLE
[PAN-2829] Account versioning

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Initial versioning code.  Main hook is that setCode() in MutableAccount now
requires a version (currently a long for performance reasons).

Reference tests all pass (hence the WorldStateMock if block) and
GenesisStateTest has an explicit version test now.

## Fixed Issue(s)
PAN-2829

### CODE REVIEW

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java [hunk 1]
```diff
@@ -89,15 +90,19 @@ public void shouldReturnEmptyCodeHashWhenPrecompileHasBalance() {
   @Test
   public void shouldGetHashOfAccountCodeWhenCodeIsPresent() {
     final BytesValue code = BytesValue.fromHexString("0xabcdef");
-    worldStateUpdater.getOrCreate(REQUESTED_ADDRESS).setCode(code, Account.DEFAULT_VERSION);
+    MutableAccount account = worldStateUpdater.getOrCreate(REQUESTED_ADDRESS);
```
Comment (abdelhamidbakhta) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java]: (nit) final

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java [hunk 2]
```diff
@@ -89,15 +90,19 @@ public void shouldReturnEmptyCodeHashWhenPrecompileHasBalance() {
...
+    account.setVersion(Account.DEFAULT_VERSION);
     assertThat(executeOperation(REQUESTED_ADDRESS)).isEqualTo(Hash.hash(code));
   }
 
   @Test
   public void shouldZeroOutLeftMostBitsToGetAddress() {
     // If EXTCODEHASH of A is X, then EXTCODEHASH of A + 2**160 is X.
     final BytesValue code = BytesValue.fromHexString("0xabcdef");
-    worldStateUpdater.getOrCreate(REQUESTED_ADDRESS).setCode(code, Account.DEFAULT_VERSION);
+    MutableAccount account = worldStateUpdater.getOrCreate(REQUESTED_ADDRESS);
```
Comment (abdelhamidbakhta) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java]: (nit) final

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/DefaultMutableWorldStateTest.java [hunk 3]
```diff
@@ -482,8 +481,8 @@ public void replaceAccountCode() {
     final WorldUpdater updater = worldState.updater();
     final MutableAccount account = updater.createAccount(ADDRESS);
     account.setBalance(Wei.of(100000));
-    account.setCode(BytesValue.of(1, 2, 3), Account.DEFAULT_VERSION);
-    account.setCode(BytesValue.of(3, 2, 1), Account.DEFAULT_VERSION);
+    account.setCode(BytesValue.of(1, 2, 3));
```
Comment (abdelhamidbakhta) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/DefaultMutableWorldStateTest.java]: (Discussion) The version is not set in the test, should we set explicitly the version and check it ?

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java (hunk 2 above)
Comment (shemnon) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java]: done

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/vm/operations/ExtCodeHashOperationTest.java]: done

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/DefaultMutableWorldStateTest.java (hunk 3 above)
Comment (shemnon) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/DefaultMutableWorldStateTest.java]: done

**Pull request merged**